File: rust/3612_process_str/process_str.rs
struct Solution;

impl Solution {
    pub fn process_str(s: String) -> String {
        let mut result: Vec<char> = Vec::new();
        for ch in s.chars() {
            match ch {
                '*' => {
                    result.pop();
                }
                '#' => {
                    let mut copy = result.clone();
                    result.append(&mut copy);
                }
                '%' => {
                    result.reverse();
                }
                c => result.push(c),
            }
        }
        result.into_iter().collect()
    }
}

fn main() {
    let tests = vec![("a#b%*", "ba"), ("z*#", "")];

    for (s, expected) in tests {
        assert_eq!(Solution::process_str(s.to_string()), expected);
    }
}

File: cpp/3614_process_str/process_str.cpp
#include <string>
#include <tuple>
#include <vector>

using namespace std;

class Solution {
 public:
  char processStr(string s, long long k) {
    long long len = 0;
    for (auto c : s) {
      if (c == '*') {
        if (len) {
          len--;
        }
      } else if (c == '#') {
        len *= 2;
      } else if (c == '%') {
        continue;
      } else {
        len++;
      }
    }
    if (k + 1 > len) {
      return '.';
    }
    for (int i = s.size() - 1; i >= 0; i--) {
      if (s[i] == '*') {
        len++;
      } else if (s[i] == '#') {
        if (k + 1 > (len + 1) / 2) {
          k -= len / 2;
        }
        len = (len + 1) / 2;
      } else if (s[i] == '%') {
        k = len - k - 1;
      } else {
        if (k + 1 == len) {
          return s[i];
        } else {
          len--;
        }
      }
    }
    return '.';
  }
};

int main() {
  vector<tuple<string, long long, char>> tests{
      {"a#b%*", 1, 'a'},
      {"cd%#*#", 3, 'd'},
      {"z*#", 0, '.'},
  };

  for (auto [s, k, expected] : tests) {
    auto result = Solution().processStr(s, k);
    if (result != expected) {
      printf("failed: %s, %lld, %c, %c\n", s.c_str(), k, expected, result);
    }
  }
}